README.md
# Mojolicious::Plugin::FormFieldsFromJSON::Date

[Mojolicious::Plugin::FormFieldsFromJSON::Date](Mojolicious::Plugin::FormFieldsFromJSON::Dat://metacpan.org/pod/Mojolicious::Plugin::FormFieldsFromJSON::Date)
is an extension to the [Mojolicious](https://metacpan.org/pod/Mojolicious)
plugin
[Mojolicious::Plugin::FormFieldsFromJSON](https://metacpan.org/pod/Mojolicious::Plugin::FormFieldsFromJSON)
for the creation of date fields.

## Usage

```
# Mojolicious
$self->plugin('FormFieldsFromJSON::Date');
$self->plugin('FormFieldsFromJSON' => {
  types => {
    date => 1,
  },
});

# Mojolicious::Lite
plugin 'FormFieldsFromJSON::Date';
plugin 'FormFieldsFromJSON' => {
  types => {
    date => 1,
  },
};
```

## Installation

The simplest way to install the module is via `cpanm`:

```
$ cpanm Mojolicious::Plugin::FormFieldsFromJSON::Date
```

Alternatively, you could download the distribution tarball and install it
in the traditional manner by using `make`:

```
$ tar -xvzf Mojolicious-Plugin-FormFieldsFromJSON-Date-<version>.tar.gz
$ cd Mojolicious-Plugin-FormFieldsFromJSON-Date-<version>
$ perl Makefile.PL
$ make
$ make test
$ make install
```

## Development

The distribution is contained in a Git repository, so simply clone the
repository

```
$ git clone git@github.com:reneeb/Mojolicious-Plugin-FormFieldsFromJSON-Date.git
```

and change into the newly-created directory.

```
$ cd Mojolicious-Plugin-FormFieldsFromJSON-Date
```

The project uses [`Dist::Zilla`](https://metacpan.org/pod/Dist::Zilla) to
build the distribution, hence this will need to be installed before
continuing:

```
$ cpanm Dist::Zilla
```

To install the required prequisite packages, run the following set of
commands:

```
$ dzil authordeps --missing | cpanm
$ dzil listdeps --author --missing | cpanm
```

The distribution can be tested like so:

```
$ dzil test
```

To run the full set of tests (including author and release-process tests),
add the `--author` and `--release` options:

```
$ dzil test --author --release
```


Changes
Release history for Mojolicious::Plugin::FormFieldsFromJSON::Date
=================================================================

  [BUGFIX]
  * pass startyear to year dropdown generation

  [TESTS]
  * shown days an months aren't formatted, so fix the tests

0.02 2014-12-31

  [TESTS]
  * The tests worked only on Dec 29th, 2014. This is fixed now!
  * Added tests for default values (defined in field config, via request, passed in method call and fallback)

0.01 2014-12-29

  [INIT]
  * allow additional type "date" for Mojolicious::Plugin::FormFieldsFromJSON


lib/Mojolicious/Plugin/FormFieldsFromJSON/Date.pm
package Mojolicious::Plugin::FormFieldsFromJSON::Date;
use Mojo::Base 'Mojolicious::Plugin';

our $VERSION = '0.03';

use Time::Piece;

sub register {
    my ($self, $app) = @_;

    1;
}

sub Mojolicious::Plugin::FormFieldsFromJSON::_date {
    my ($self, $c, $field, %params) = @_;

    my $name         = $field->{name} // $field->{label} // '';
    my $id           = $field->{id} // $name;
    my %local_params = %{ $params{$name} || {} };

    my $requested = ( $c->param( $name . '-day' ) ) ?
       ( sprintf "%04d-%02d-%02d",
           $c->param( $name . '-year' ),
           $c->param( $name . '-month' ),
           $c->param( $name . '-day' ),
       ) :
       "";

    my $now      = localtime;
    my $default  = $local_params{data} || $requested || $field->{data} || $now->ymd;
    my $selected = Time::Piece->strptime( $default, '%Y-%m-%d' );

    my $day = _day_dropdown(
        $c,
        %{ $field->{attributes} || {} },
        %local_params,
        day  => $selected->mday,
        id   => $id . '-day',
        name => $name . '-day',
    );

    my $month = _month_dropdown(
        $c,
        %{ $field->{attributes} || {} },
        %local_params,
        month => $selected->mon,
        id    => $id . '-month',
        name  => $name . '-month',
    );

    my $year = _year_dropdown(
        $c,
        startyear => $field->{startyear},
        %{ $field->{attributes} || {} },
        %local_params,
        year => $selected->year,
        id   => $id . '-year',
        name => $name . '-year',
    );

    my $format = $params{format} || $field->{format} || '%Y-%m-%d';
    $format =~ s/-//g;
    $format =~ s/\%Y/$year/g;
    $format =~ s/\%m/$month/g;
    $format =~ s/\%d/$day/g;

    return $format;
}

sub _day_dropdown {
    my ($c, %params) = @_;

    my @days = map{
        my %opts = ( $_ == $params{day} ) ?
            ('selected' => 'selected') :
            ();
        [ $_ => sprintf("%02d", $_) , %opts ]
    }(1 .. 31);

    $c->param( $params{name}, '' );

    my $select = $c->select_field(
        $params{name},
        \@days,
        id => $params{id},
        %{ $params{attrs} || {} },
    );
}

sub _month_dropdown {
    my ($c, %params) = @_;

    my @months = map{
        my %opts = ( $_ == $params{month} ) ?
            ('selected' => 'selected') :
            ();
        [ $_ => sprintf("%02d", $_), %opts ]
    }(1 .. 12);

    $c->param( $params{name}, '' );

    my $select = $c->select_field(
        $params{name},
        \@months,
        id => $params{id},
        %{ $params{attrs} || {} },
    );
}

sub _year_dropdown {
    my ($c, %params) = @_;

    my $now   = localtime;
    my $base  = ( $params{startyear} // $now->year );
    my $start = $base - ( $params{past} // 1 );
    my $stop  = $base + ( $params{future} // 5 );

    $c->param( $params{name}, '' );

    my $has_year;
    my @years = map{
        $has_year++ if $_ == $params{year};
        my %opts = ( $_ == $params{year} ) ?
            ('selected' => 'selected') :
            ();
        [ $_ => $_, %opts ]
    }($start .. $stop);

    if ( !$has_year ) {
        unshift @years, [ $params{year} => $params{year}, { selected => 'selected' } ];
    }

    my $select = $c->select_field(
        $params{name},
        \@years,
        id => $params{id},
        %{ $params{attrs} || {} },
    );
}

1;

__END__

=encoding utf8

=head1 NAME

Mojolicious::Plugin::FormFieldsFromJSON::Date - Mojolicious Plugin

=head1 SYNOPSIS

  # Mojolicious
  $self->plugin('FormFieldsFromJSON::Date');
  $self->plugin('FormFieldsFromJSON' => {
    types => {
      date => 1,
    },
  });

  # Mojolicious::Lite
  plugin 'FormFieldsFromJSON::Date';
  plugin 'FormFieldsFromJSON' => {
    types => {
      date => 1,
    },
  };

=head1 DESCRIPTION

L<Mojolicious::Plugin::FormFieldsFromJSON::Date> is an extension for
L<Mojolicious::Plugin::FormFieldsFromJSON> to create date fields.

=head1 CONFIGURATION

You can configure several things for the date fields:

=over 4

=item * selected

The date selected. Must be in ISO format, e.g.

  2014-12-29

Default: "today"

=item * format

Defines the order of the dropdowns.

  %Y-%m-%d

is the default and this will create the "year"-dropdown first, then
the "month"-dropdown and finally the "day"-dropdown.

=item * future

Defines how many years in the future should be available. (Default: 5)

=item * past

Defines how many years in the past should be available. (Default: 1)

=back

=head1 DATE FIELDS

The date fields is a collection of dropdowns to select a date.

=head2 A simple date field

This is an example for a very simple date field

Configuration:

  [
    {
      "name":"start",
      "future": 2,
      "past": 0,
      "format":"%d-%m-%Y"
    }
  ]

=head1 SEE ALSO

L<Mojolicious>, L<Mojolicious::Guides>, L<http://mojolicio.us>,
L<Mojolicious::Plugin::FormFieldsFromJSON>.

=cut

